File: Java/SpectableJavaExamplesTest/src/main/java/production/SimpleJson.java
package production;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The compact JSON-like text used by the Json specification: names are not
 * quoted, values are, and whitespace between tokens is insignificant.
 *
 *     {anInt:"1",aString:"B"}
 *     [{anInt:"1",aString:"B"},{anInt:"2",aString:"C"}]
 *
 * Field order is preserved, so a canonical form can be compared directly.
 */
public class SimpleJson {

    // -----------------------------------------------------------------
    // Writing
    // -----------------------------------------------------------------

    public static String toObject(Map<String, String> fields) {
        StringBuilder sb = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, String> field : fields.entrySet()) {
            if (!first) sb.append(',');
            first = false;
            sb.append(field.getKey()).append(':');
            appendQuoted(sb, field.getValue());
        }
        return sb.append('}').toString();
    }

    public static String toArray(List<? extends Map<String, String>> rows) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < rows.size(); i++) {
            if (i > 0) sb.append(',');
            sb.append(toObject(rows.get(i)));
        }
        return sb.append(']').toString();
    }

    private static void appendQuoted(StringBuilder sb, String value) {
        sb.append('"');
        if (value != null) {
            for (int i = 0; i < value.length(); i++) {
                char c = value.charAt(i);
                if (c == '"' || c == '\\') sb.append('\\');
                sb.append(c);
            }
        }
        sb.append('"');
    }

    // -----------------------------------------------------------------
    // Reading
    // -----------------------------------------------------------------

    public static LinkedHashMap<String, String> parseObject(String text) {
        Cursor c = new Cursor(text);
        LinkedHashMap<String, String> fields = readObject(c);
        c.skipWhitespace();
        if (!c.atEnd()) throw new IllegalArgumentException("Unexpected text after object at " + c.pos());
        return fields;
    }

    public static List<LinkedHashMap<String, String>> parseArray(String text) {
        Cursor c = new Cursor(text);
        c.skipWhitespace();
        c.expect('[');
        List<LinkedHashMap<String, String>> rows = new ArrayList<>();
        c.skipWhitespace();
        if (c.peek() == ']') {
            c.next();
        } else {
            while (true) {
                rows.add(readObject(c));
                c.skipWhitespace();
                char d = c.next();
                if (d == ',') continue;
                if (d == ']') break;
                throw new IllegalArgumentException("Expected ',' or ']' at " + c.pos());
            }
        }
        c.skipWhitespace();
        if (!c.atEnd()) throw new IllegalArgumentException("Unexpected text after array at " + c.pos());
        return rows;
    }

    /**
     * Remove whitespace that sits between tokens, leaving a plain string that
     * can be compared to another one directly. Whitespace inside a quoted
     * value is part of the value and is kept.
     */
    public static String withoutWhitespace(String text) {
        if (text == null) return "";
        StringBuilder sb = new StringBuilder(text.length());
        boolean inQuotes = false;
        for (int i = 0; i < text.length(); i++) {
            char c = text.charAt(i);
            if (inQuotes) {
                sb.append(c);
                if (c == '\\' && i + 1 < text.length()) {
                    sb.append(text.charAt(i + 1));
                    i++;
                } else if (c == '"') {
                    inQuotes = false;
                }
            } else if (c == '"') {
                inQuotes = true;
                sb.append(c);
            } else if (!Character.isWhitespace(c)) {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    private static LinkedHashMap<String, String> readObject(Cursor c) {
        c.skipWhitespace();
        c.expect('{');
        LinkedHashMap<String, String> fields = new LinkedHashMap<>();
        c.skipWhitespace();
        if (c.peek() == '}') { c.next(); return fields; }

        while (true) {
            c.skipWhitespace();
            String name = readName(c);
            c.skipWhitespace();
            c.expect(':');
            c.skipWhitespace();
            fields.put(name, readValue(c));
            c.skipWhitespace();
            char d = c.next();
            if (d == ',') continue;
            if (d == '}') return fields;
            throw new IllegalArgumentException("Expected ',' or '}' at " + c.pos());
        }
    }

    /** A name is bare text up to the colon, or a quoted string. */
    private static String readName(Cursor c) {
        if (c.peek() == '"') return readQuoted(c);
        StringBuilder sb = new StringBuilder();
        while (!c.atEnd() && c.peek() != ':') sb.append(c.next());
        return sb.toString().trim();
    }

    /** A value is a quoted string, or bare text up to the next ',' or '}'. */
    private static String readValue(Cursor c) {
        if (c.peek() == '"') return readQuoted(c);
        StringBuilder sb = new StringBuilder();
        while (!c.atEnd() && c.peek() != ',' && c.peek() != '}') sb.append(c.next());
        return sb.toString().trim();
    }

    private static String readQuoted(Cursor c) {
        c.expect('"');
        StringBuilder sb = new StringBuilder();
        while (true) {
            if (c.atEnd()) throw new IllegalArgumentException("Unterminated string at " + c.pos());
            char ch = c.next();
            if (ch == '"') return sb.toString();
            if (ch == '\\' && !c.atEnd()) ch = c.next();
            sb.append(ch);
        }
    }

    private static final class Cursor {
        private final String text;
        private int i = 0;

        Cursor(String text) { this.text = text == null ? "" : text; }

        int pos()       { return i; }
        boolean atEnd() { return i >= text.length(); }
        char peek()     { return atEnd() ? '\0' : text.charAt(i); }
        char next()     { return atEnd() ? '\0' : text.charAt(i++); }

        void skipWhitespace() {
            while (!atEnd() && Character.isWhitespace(text.charAt(i))) i++;
        }

        void expect(char expected) {
            skipWhitespace();
            if (next() != expected)
                throw new IllegalArgumentException("Expected '" + expected + "' at " + i);
        }
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/production/SimpleJson.cs
namespace production
{
    using System;
    using System.Collections.Generic;
    using System.Text;

    /// <summary>
    /// The compact JSON-like text used by the Json specification: names are not
    /// quoted, values are, and whitespace between tokens is insignificant.
    ///
    ///     {anInt:"1",aString:"B"}
    ///     [{anInt:"1",aString:"B"},{anInt:"2",aString:"C"}]
    ///
    /// Field order is preserved, so a canonical form can be compared directly.
    /// </summary>
    public static class SimpleJson
    {
        // -----------------------------------------------------------------
        // Writing
        // -----------------------------------------------------------------

        public static string ToObject(IEnumerable<KeyValuePair<string, string>> fields)
        {
            var sb = new StringBuilder("{");
            bool first = true;
            foreach (var field in fields)
            {
                if (!first) sb.Append(',');
                first = false;
                sb.Append(field.Key).Append(':');
                AppendQuoted(sb, field.Value);
            }
            return sb.Append('}').ToString();
        }

        public static string ToArray(List<Dictionary<string, string>> rows)
        {
            var sb = new StringBuilder("[");
            for (int i = 0; i < rows.Count; i++)
            {
                if (i > 0) sb.Append(',');
                sb.Append(ToObject(rows[i]));
            }
            return sb.Append(']').ToString();
        }

        private static void AppendQuoted(StringBuilder sb, string? value)
        {
            sb.Append('"');
            if (value != null)
            {
                foreach (char c in value)
                {
                    if (c == '"' || c == '\\') sb.Append('\\');
                    sb.Append(c);
                }
            }
            sb.Append('"');
        }

        // -----------------------------------------------------------------
        // Reading
        // -----------------------------------------------------------------

        public static Dictionary<string, string> ParseObject(string text)
        {
            var c = new Cursor(text);
            var fields = ReadObject(c);
            c.SkipWhitespace();
            if (!c.AtEnd) throw new FormatException($"Unexpected text after object at {c.Pos}");
            return fields;
        }

        public static List<Dictionary<string, string>> ParseArray(string text)
        {
            var c = new Cursor(text);
            c.SkipWhitespace();
            c.Expect('[');
            var rows = new List<Dictionary<string, string>>();
            c.SkipWhitespace();
            if (c.Peek() == ']')
            {
                c.Next();
            }
            else
            {
                while (true)
                {
                    rows.Add(ReadObject(c));
                    c.SkipWhitespace();
                    char d = c.Next();
                    if (d == ',') continue;
                    if (d == ']') break;
                    throw new FormatException($"Expected ',' or ']' at {c.Pos}");
                }
            }
            c.SkipWhitespace();
            if (!c.AtEnd) throw new FormatException($"Unexpected text after array at {c.Pos}");
            return rows;
        }

        /// <summary>
        /// Remove whitespace that sits between tokens, leaving a plain string that
        /// can be compared to another one directly. Whitespace inside a quoted
        /// value is part of the value and is kept.
        /// </summary>
        public static string WithoutWhitespace(string? text)
        {
            if (text == null) return string.Empty;
            var sb = new StringBuilder(text.Length);
            bool inQuotes = false;
            for (int i = 0; i < text.Length; i++)
            {
                char c = text[i];
                if (inQuotes)
                {
                    sb.Append(c);
                    if (c == '\\' && i + 1 < text.Length)
                    {
                        sb.Append(text[i + 1]);
                        i++;
                    }
                    else if (c == '"')
                    {
                        inQuotes = false;
                    }
                }
                else if (c == '"')
                {
                    inQuotes = true;
                    sb.Append(c);
                }
                else if (!char.IsWhiteSpace(c))
                {
                    sb.Append(c);
                }
            }
            return sb.ToString();
        }

        // -----------------------------------------------------------------

        private static Dictionary<string, string> ReadObject(Cursor c)
        {
            c.SkipWhitespace();
            c.Expect('{');
            var fields = new Dictionary<string, string>();
            c.SkipWhitespace();
            if (c.Peek() == '}') { c.Next(); return fields; }

            while (true)
            {
                c.SkipWhitespace();
                string name = ReadName(c);
                c.SkipWhitespace();
                c.Expect(':');
                c.SkipWhitespace();
                fields[name] = ReadValue(c);
                c.SkipWhitespace();
                char d = c.Next();
                if (d == ',') continue;
                if (d == '}') return fields;
                throw new FormatException($"Expected ',' or '}}' at {c.Pos}");
            }
        }

        /// <summary>A name is bare text up to the colon, or a quoted string.</summary>
        private static string ReadName(Cursor c)
        {
            if (c.Peek() == '"') return ReadQuoted(c);
            var sb = new StringBuilder();
            while (!c.AtEnd && c.Peek() != ':') sb.Append(c.Next());
            return sb.ToString().Trim();
        }

        /// <summary>A value is a quoted string, or bare text up to the next ',' or '}'.</summary>
        private static string ReadValue(Cursor c)
        {
            if (c.Peek() == '"') return ReadQuoted(c);
            var sb = new StringBuilder();
            while (!c.AtEnd && c.Peek() != ',' && c.Peek() != '}') sb.Append(c.Next());
            return sb.ToString().Trim();
        }

        private static string ReadQuoted(Cursor c)
        {
            c.Expect('"');
            var sb = new StringBuilder();
            while (true)
            {
                if (c.AtEnd) throw new FormatException($"Unterminated string at {c.Pos}");
                char ch = c.Next();
                if (ch == '"') return sb.ToString();
                if (ch == '\\' && !c.AtEnd) ch = c.Next();
                sb.Append(ch);
            }
        }

        private sealed class Cursor
        {
            private readonly string _text;
            private int _i;

            public Cursor(string? text) { _text = text ?? string.Empty; }

            public int Pos => _i;
            public bool AtEnd => _i >= _text.Length;
            public char Peek() => AtEnd ? '\0' : _text[_i];
            public char Next() => AtEnd ? '\0' : _text[_i++];

            public void SkipWhitespace()
            {
                while (!AtEnd && char.IsWhiteSpace(_text[_i])) _i++;
            }

            public void Expect(char expected)
            {
                SkipWhitespace();
                if (Next() != expected)
                    throw new FormatException($"Expected '{expected}' at {_i}");
            }
        }
    }
}
